### PR TITLE
parser: improved SELECT parenthesization

### PR DESCRIPTION
**Pull request by sougou**
@bbeaudreault @acharis 

This change allows the following extensions to the SQL syntax:
1. UNION statements can be followed by ORDER BY, LIMIT and LOCK clauses.
2. SELECT statements that are part of a UNION can be parenthesized.
3. INSERT...SELECT allows SELECT to be parenthesized.

The main purpose of allowing parenthesis for selects was to
support the following two constructs: (SELECT...) UNION (SELECT...)
and INSERT INTO a (SELECT...).

The initial approach was to allow parenthesizing of all SELECT
statements. However, just adding that as an additional rule
created conflicts. In particular, `((SELECT * FROM t))` could
be seen as a subquery with a parenthesized select, or a row
tuple with a subquery. We've avoided this conflict by allowing
parenthesis only for selects that are part of a union.

For UNION constructs, we go one step further and allow
ORDER BY and LIMIT clauses for unions which we didn't allow
before. The tabletserver planbuilder has also been extended
accordingly. VTGate support of unions will also need to be extended.

For the inserts, it's only a syntax improvement. So, nothing else
needs to change. The second change resolves the shift/reduce
conflict introduced by the first.

If necessary, we could allow the top level statement to be
parenthesized, but that has no practical value. So, we can
skip that for now.

### CODE REVIEW

File: go/vt/sqlparser/Makefile [hunk 1]
```diff
@@ -5,7 +5,7 @@
 MAKEFLAGS = -s
 
 sql.go: sql.y
-	go tool yacc -o sql.go sql.y
+	goyacc -o sql.go sql.y
```
Comment (bbeaudreault) [inline go/vt/sqlparser/Makefile]: I don't seem to have `goyacc`, but `go tool yacc` works for me. Am I missing something, or was there a reason for this change?

**Comment by bbeaudreault** [comment]
Other than that one question about the Makefile, this looks great. I think the compromise is totally fine for now. I can see how this would be a pain to figure out in the grammar, but now that it's there I don't think it makes it too messy.  Thanks for doing this!

Test failures look to be due to a 404 downloading mariadb. Master is failing as well.

File: go/vt/sqlparser/Makefile (hunk 1 above)
Comment (sougou) [inline go/vt/sqlparser/Makefile]: This is because I'm on go1.8. I'll revert this.

**Comment by sougou** [comment]
@bbeaudreault @acharis I've added support for INSERT. If you LGTM, I'll use it as approval to submit this change.

**Comment by bbeaudreault** [comment]
This LGTM

**Comment by acharis** [comment]
agreed that this is a good compromise. nice job avoiding a larger refactor!  : )
it would be good to see a green travis. i don't want a failure from one problem to mask a failure from another. otherwise LGTM.

**Comment by bbeaudreault** [comment]
I think the failure could be fixed by rebasing. Mariadb URL changed and was recently fixed in master.

**Comment by sougou** [comment]
LGTM
I ran the tests locally and they all pass.

**Pull request merged**